Index Resolver Values

shrugs opened the pull request:
beginning of https://github.com/namehash/ensnode/issues/723

- adds a set of `ext_` prefixed entities to track resolver records
  - this is safe for subgraph compatibility at the api layer because ENSIndexer will now serve a superset of the subgraph entities.
  - i don't think it's currently trivial to conditionally include/exclude the `ext_` entities from the schema (and the autogenerated api) based on whether this runtime flag is enabled, so if `INDEX_RESOLVER_RECORDS` is false, these records will just not exist and any api calls will return empty array or null
  - currently just represents address records by coinType, text records by key, and `name` under `records` — can add the others if necessary but they seem not useful at the current moment
  - access via `domain(id: "") { resolver { addressRecords textRecords } }` (can't use `texts` because conflicts with existing `texts` coinType array from subgraph)
- adds a derived `isSubgraphCompatible` flag to the `ENSIndexerConfig` which is derived from the other inputs
  - it's more or less just an internal representation of whether or not the shared behaviors should operate with subgraph compatibility in mind or not
  - i don't _love_ that it's in ENSIndexerConfig but i think it's relatiely aligned, as ENSIndexerConfig is, specifically, an internal configuration object. it's the `ENSIndexerEnvironment` that's user-facing and intended to be used. lmk if you have any other ideas. i didn't like the idea of just adding a global based on `config`, nor did i love a `ENSIndexerInternalConfig` that's built from `ENSIndexerConfig`
- refactors all of the id generation to use the new `config.isSubgraphCompatible` flag
  - this resolves the resolver reference issue i noticed [in slack here](https://namehash.slack.com/archives/C086Z6FNBHN/p1748018943438899)
  - the best way to review this change is via [this commit](https://github.com/namehash/ensnode/commit/e25650717bace6b352e35639336fe04939346bfc) in isolation with hide whitespace turned on
- added tests for the new `isSubgraphCompatible` flag and refactored the `ids.test.ts` tests for the new pattern
- adds pretty logging of the config at startup for easier debugging 
- refactors the config.schema to be more zod-like [in this commit](https://github.com/namehash/ensnode/pull/746/commits/90f8fa95b122b2ad29a7d548efdf351e0bc63655)
---

if you review this pr all at once don't forget to hide whitespace diffs

---

<img width="1625" alt="Screenshot 2025-05-23 at 13 56 28" src="https://github.com/user-attachments/assets/c0663e48-7696-438c-a1c5-196bebb6e42c" />

<img width="1649" alt="image" src="https://github.com/user-attachments/assets/671e80d3-2370-4b97-afab-81fbfb360e99" />

tk-o commented on apps/ensindexer/src/handlers/NameWrapper.ts:
```diff
@@ -3,9 +3,9 @@ import schema from "ponder:schema";
+import { type Node, uint256ToHex32 } from "@ensnode/utils";
 
-import { makeSharedEventValues, upsertAccount } from "@/lib/db-helpers";
+import { sharedEventValues, upsertAccount } from "@/lib/db-helpers";
```
`sharedEventValues` is not used in this file. Do we need to use that import? Otherwise, we should drop it.

shrugs replied:
we still use it, it's just hidden in the diff because those lines didn't change — previously we were constructing this function using the `make*` factory but now it's just the function `sharedEventValues` so all of those events that were logging are still using the same function. applies to similar comments below

tk-o commented on apps/ensindexer/src/handlers/Registrar.ts:
```diff
@@ -10,7 +10,7 @@ import {
 } from "@ensnode/utils";
 
-import { makeSharedEventValues, upsertAccount, upsertRegistration } from "@/lib/db-helpers";
+import { sharedEventValues, upsertAccount, upsertRegistration } from "@/lib/db-helpers";
```
I can't see `sharedEventValues` function being called anywhere in this file. Should we drop the import, or use the function somehow?

tk-o commented on apps/ensindexer/src/handlers/Registrar.ts:
```diff
@@ -59,7 +58,7 @@ export const makeRegistrarHandlers = ({
     await context.db
       .update(schema.registration, {
-        id: makeRegistrationId(pluginName, labelHash, node),
+        id: makeRegistrationId(labelHash, node),
```
Nice one on using the config context to drop explicit `pluginName` dependency 👍

tk-o commented on apps/ensindexer/ponder.config.ts:
```diff
@@ -31,7 +32,10 @@ export type MergedPonderConfig = MergedTypes<(typeof ALL_PLUGINS)[number]["confi
-  indexingBehaviorDependencies: Pick<ENSIndexerConfig, "healReverseAddresses">;
+  indexingBehaviorDependencies: Pick<
+    ENSIndexerConfig,
+    "healReverseAddresses" | "indexResolverRecords"
```
👍

tk-o commented on apps/ensindexer/src/config/config.schema.ts:
```diff
@@ -12,177 +12,173 @@ import {
 
-const parseBlockNumber = (envVarKey: string) =>
+// parses an env string bool with strict requirement of 'true' or 'false'
+const makeEnvStringBoolSchema = (envVarKey: string) =>
+  z
+    .string()
+    .pipe(
+      z.enum(["true", "false"], {
+        error: `${envVarKey} must be 'true' or 'false'.`,
+      }),
+    )
+    .transform((val) => val === "true");
```
Nice helper schema 👍

tk-o commented on apps/ensindexer/src/config/config.schema.ts:
```diff
@@ -12,177 +12,173 @@ import {
+    databaseUrl: DatabaseUrlSchema,
   })
+  // inject ENSIndexerConfig.subgraphCompatibility
+  .transform(derive_subgraphCompatibility)
```
Cool pattern! It makes me keen to learn Zod v4 API 🤓

lightwalker-eth reviewed (commented):
@shrugs Hey super to see this PR! Great work and excited for the future possibilities here 🫡 Reviewed and shared some questions with feedback.

Also will be great to include updates to the following in this PR:

1. https://github.com/namehash/ensnode/blob/main/apps/ensindexer/.env.local.example
2. Suggest we expand the docs for the "Purpose" of the ENSNode 'Alpha' deployment here: https://ensnode.io/docs/usage/hosted-ensnode-instances/#available-ensnode-instances to identify how we are also beginning other experimental work for ENSv2 indexing such as ENS Protocol Acceleration.
3. Could you please review our docs and identify any areas that might be helpful to update or expand upon?
4. How about the GraphQL schema docs?

shrugs commented on apps/ensindexer/src/plugins/shared/Resolver.ts:
```diff
@@ -11,58 +10,41 @@ import { ENSIndexerPluginHandlerArgs } from "@/lib/plugin-helpers";
   // NOTE: this DNSRecordChanged ABI spec with the included `ttl` parameter is specific to
   // 3DNS' Resolver implementation, but we include it here for theoretical completeness, were a
-  // Resolver indexed by these shared handlers to emit this event.
+  // Resolver indexed by these shared rh.handlers to emit this event.
```
vestige of previous work in the pr

shrugs commented on apps/ensindexer/test/config.test.ts:
```diff
@@ -132,17 +132,17 @@ describe("config", () => {
     it("throws an error if ENSNODE_PUBLIC_URL is not a valid URL", async () => {
       vi.stubEnv("ENSNODE_PUBLIC_URL", "invalid url");
-      await expect(getConfig()).rejects.toThrow(/ENSNODE_PUBLIC_URL must be a valid URL string/i);
+      await expect(getConfig()).rejects.toThrow(/must be a valid URL string/i);
```
sure thing!

shrugs commented:
this has been updated:
- i've folded the `ResolverRecords` entity into `Resolver` and just added the `name` property to it — it seems simplest.
- added a bunch of documentation related to the Resolver records schema that should help clear up any confusion in that file.

shrugs commented:
also confirmed subgraph compatible

tk-o commented on apps/ensindexer/src/config/config.schema.ts:
```diff
@@ -12,177 +12,179 @@ import {
+    databaseUrl: DatabaseUrlSchema,
   })
+  // inject ENSIndexerConfig.isSubgraphCompatible
+  .transform(derive_isSubgraphCompatible)
```
Thinking about deriving the `ensDeployment` config prop. Right now, whenever we need to access the current ENS Deployment object, we'd import `getENSDeployment` function and pass the `config.ensDeploymentChain` as a param:
![image](https://github.com/user-attachments/assets/0f92a368-433c-459b-99e3-6d4117434bb0)

I propose we derive `ensDeployment` property and include it on `ENSIndexerSchema`

```suggestion
  .transform(derive_isSubgraphCompatible)
  .transform(derive_ensDeployment)
```

Where `derive_ensDeployment` is defined as the following:
```ts
const derive_ensDeployment = (
  config: Omit<ENSIndexerConfig, "ensDeployment">,
): ENSIndexerConfig {
  return {
    ...config,
    ensDeployment: getENSDeployment(config.ensDeploymentChain),
  };

tk-o replied:
If we feel the idea is OK, but it needs a follow up PR, I'm happy to create a GH issue ;) 

Also, thinking of deriving active plugins list in a similar fashion, and later using this list in `ponder.config.ts` to actually activate relevant plugins.

shrugs replied:
actually, splitting this into a separate issue

tk-o commented on apps/ensindexer/src/lib/lib-config.ts:
```diff
@@ -43,3 +43,23 @@ export function getRpcConfigsFromEnv(): Record<number, RpcConfigEnvironment> {
 }
+
+export function prettyPrintConfig(config: ENSIndexerConfig) {
+  return JSON.stringify(
+    {
+      ...config,
+      databaseUrl: "*******",
+      rpcConfigs: Object.fromEntries(
+        Object.entries(config.rpcConfigs).map(([chainId, rpcConfig]) => [
+          chainId,
+          {
+            ...rpcConfig,
+            url: "*******",
+          },
+        ]),
+      ),
+    } as ENSIndexerConfig,
+    null,
+    2,
+  );
```
Feel like a nice candidate for the `toPublic` method we spoke about some other time: https://github.com/namehash/ensnode/issues/521#issuecomment-2802662311

I'll take a note to come back to this function once ENSNode SDK comes to life 🚀

tk-o reviewed (approved):
I left couple of comments, but nothing critical. Merge when ready 🚀

lightwalker-eth commented on packages/ensnode-schema/src/resolver-records.schema.ts:
```diff
@@ -1,24 +1,73 @@
+ *   - key
+ *   - value
+ *
+ * This results in the following query patterns (ponder graphql pseudo code) to look up the records
```
My attention now is focused on the word "records" in this sentence. I think it will help us to introduce more terminology to help us communicate about these ideas with additional clarity and precision:

* "Lite Indexed Records": The resolver records (keys and selected subset of values) indexed by the ENS Subgraph.
* "Extended Indexed Records": The resolver records (keys and selected subset of values) indexed by ENSNode when we have this optional feature flag activated.
* "Effective Protocol Records": The resolver records that would actually be returned for a node through dynamic lookups using the ENS Protocol.

This terminology can help us in a number of ways. For example:
1. Identify how great caution needs to be taken when reading either "Lite Indexed Records" or "Extended Indexed Records" as these don't recognize ENSIP-10 or CCIP-Read and therefore may differ from "Effective Protocol Records".
2. Explicitly give name to the distinction between "Lite Indexed Records" and "Extended Indexed Records"
3. Explicitly give name to the distinction between "Extended Indexed Records" and "Effective Protocol Records"

shrugs replied:
i'm less enthused about this terminology approach and would instead prefer we effectively say "resolving records for a node via database lookups is not ENSIP-10 nor CCIP-Read compliant" or something — basically using the GraphQL API to do any sort of record resolution is a no-no. I see why the subgraph only tracks values because it's hard to police an uneducated consumer from doing this

lightwalker-eth commented on packages/ensnode-schema/src/resolver-records.schema.ts:
```diff
@@ -1,24 +1,73 @@
+ * query RecordsForNode($id: ID!) {
+ *   resolver(id: $id) {
+ *     name
+ *     addressRecords { totalCount items { coinType address } }
```
Suggest we document some special details about these addressRecords:

1. Different chains (different cointypes) encode addresses in different ways. For example for Bitcoin or Solana or etc.. the addresses will be encoded differently than Ethereum addresses.
2. I'm not sure how the above translates into our Postgres schema or GraphQL API. This might need some consideration. Not sure.
3. Additional details here: https://docs.ens.domains/ensip/9
4. https://github.com/ensdomains/ensjs/blob/main/docs/public/function.getAddressRecord.md
5. https://github.com/ensdomains/ensjs/blob/main/packages/ensjs/src/functions/public/_getAddr.ts

shrugs replied:
I think that information can be deferred to ENSIP-9, i'll add another callout there

lightwalker-eth commented on apps/ensindexer/src/lib/ids.ts:
```diff
@@ -90,11 +90,12 @@ export const makeRegistrationId = (labelHash: LabelHash, node: Node) => {
 
 /**
- * Makes a unique ID for any resolver record entity.
+ * Makes a unique ID for any resolver record entity that is keyed beyond Node
```
Suggest to also direct the reader to learn more about the distinction between resolvers in the protocol and resolver entities as indexed by the subgraph in `packages/ensnode-schema/src/subgraph.schema.ts`. Otherwise it's really easy to get confused by this stuff.

lightwalker-eth commented on apps/ensindexer/src/lib/ids.ts:
```diff
@@ -90,11 +90,12 @@ export const makeRegistrationId = (labelHash: LabelHash, node: Node) => {
  *
  * @param resolverId the id of the resolver
- * @param key any string value
+ * @param key any string value (ex: coinType.toString(), key)
```
Suggest we refine this. For example "key any string value ..." can cause confusion because both "key" and "value" are words in this sentence.

Would it work to say something like:
`@param key the unique id of the resolver record within resolverId`

Also I don't feel confident that I understand what the following represents:

`(ex: coinType.toString(), key)`

Is this saying the key will be a tuple?

shrugs commented on apps/ensindexer/src/config/config.schema.ts:
```diff
@@ -12,177 +12,179 @@ import {
-import { Address, isAddress } from "viem";
 
-const parseBlockNumber = (envVarKey: string) =>
+const chainIdSchema = z.number().int().min(1);
```
i don't think so as it's intentional that rpcConfig allows any chain to be configured, not just the ones that ensnode supports and not just the ones that are explicitly indexed — it only requires that the indexed chains have configuration at the end of the day

shrugs commented on packages/ensnode-schema/src/resolver-records.schema.ts:
```diff
@@ -1,24 +1,73 @@
+ * given Name is 'active' at any one time, and it is _that_ Resolver contract that one must query to
+ * resolve records for the Node in question.
+ *
+ * In the Subgraph schema, however a Resolver _entity_ does NOT represent a Resolver _contract_:
```
preferring the original structure of this sentence, but with an added comma after however to be grammatically correct

shrugs commented:
updates:
- chose `indexAdditionalResolverRecords` and `INDEX_ADDITIONAL_RESOLVER_RECORDS` as the config names
- because `indexAdditionalResolverRecords` results in superset of subgraph behavior, data and api, i've removed it from the `isSubgraphCompatible` derivation
- updated comments incorporating most feedback, please give them another look over

lightwalker-eth commented on packages/ensnode-schema/src/resolver-records.schema.ts:
```diff
@@ -1,66 +1,82 @@
- * NOTE: generate the (subgraph entity) Resolver ID as:
- * - (if running in subgraph-compat mode): `${resolverContractAddress.toLowerCase()}-${node}`
- * - (if not running in subgraph-compat mode): `${chainId}-${resolverContractAddress.toLowerCase()}-${node}`
+ * In this file, we extend the subgraph's Resolver _entity_ with the following:
```
```suggestion
 * In this file, we extend the subgraph's Resolver _entity_ with the following "additional resolver records":
```

lightwalker-eth commented on packages/ensnode-schema/src/resolver-records.schema.ts:
```diff
@@ -1,66 +1,82 @@
+ * Resolver (the Resolver-Node relationship)
+ *   + has one `name` record for a Node (see ENSIP-3)
+ *   + has many `addressRecords` for a Node (by coinType) (see ENSIP-9)
+ *     - ResolverAddressRecord indexes `coinType` and `address`
```
Sorry could you clarify how this should be interpreted? Not confident I follow. Suggest to write this idea out in more words and detail.

lightwalker-eth replied:
Is there any guaranteed relationship between the `addr` field being defined and the presence of a particular `addressRecords` key? No matter if that's a yes or no, this would be a nice invariant / expectation to document.

Additionally, I imagine there is a guaranteed relationship between the subgraph-compat `coinTypes` keys and the `coinType` keys defined here? Will also be nice to document this.

lightwalker-eth commented on packages/ensnode-schema/src/resolver-records.schema.ts:
```diff
@@ -1,66 +1,82 @@
+ *   + has many `addressRecords` for a Node (by coinType) (see ENSIP-9)
+ *     - ResolverAddressRecord indexes `coinType` and `address`
+ *   + has many `textRecords` for a Node (by key) (see ENSIP-5)
+ *     - ResolverTextRecord indexes `key` and `value`
```
Sorry could you clarify how this should be interpreted? Not confident I follow. Suggest to write this idea out in more words and detail.

lightwalker-eth commented on packages/ensnode-schema/src/resolver-records.schema.ts:
```diff
@@ -1,66 +1,82 @@
+ * # NOTE: generate the Resolver entity ID as:
+ * # - (if running in subgraph-compat mode): `${resolverContractAddress.toLowerCase()}-${node}`
+ * # - (if not running in subgraph-compat mode): `${chainId}-${resolverContractAddress.toLowerCase()}-${node}`
+ * query RecordsForNodeOnResolver($id: ID!) {
```
How should this example query be interpreted?

1. Is it a query for all subgraph-compat resolver records?
2. Is it a query for all "additional resolver records"?
3. Is it a query for all of both (1) and (2) listed above?
4. Is it a query for (3) listed above less fields that are redundant between (1) and (2)?

Suggest to refine how we describe and name this query to help make it explicit how it should be interpreted. Is this query case 4?

I would be interested to see an example query for all of these cases if possible 🙏 Or perhaps case 1, 2, and 4 assuming 3 provides no benefit over 4.

shrugs replied:
I'm just going to remove this example, since we don't want people querying it anyway

lightwalker-eth replied:
Is there a simple way for us to separate the idea of indexing this data from it being exposed in our GraphQL APIs? If so that would be great. I like the idea of not making this data available for direct query through our GraphQL APIs at all if possible. No worries if this is complex -- we can come back to it later too.

lightwalker-eth commented on packages/ensnode-schema/src/resolver-records.schema.ts:
````diff
@@ -1,66 +1,82 @@
  * }
  *```
+ *
+ * NOTE: The addition of these record keys and values does NOT allow the caller to confidently
````
```suggestion
 * NOTE: These "additional resolver records" do NOT allow the caller to confidently
```

Suggesting this to help maximize explicit relationships between terminology we introduce.

lightwalker-eth commented on packages/ensnode-schema/src/resolver-records.schema.ts:
````diff
@@ -1,66 +1,82 @@
  *```
+ *
+ * NOTE: The addition of these record keys and values does NOT allow the caller to confidently
+ * resolve records for names: a direct lookup of a Name's Resolver and retrieval of its values
````
```suggestion
 * resolve records for names: a direct query for the indexed values of a name's Resolver records
```

lightwalker-eth commented on packages/ensnode-schema/src/resolver-records.schema.ts:
```diff
@@ -1,66 +1,82 @@
- * for a Node within a Resolver contract:
+ * The Subgraph describes the following datamodel instead:
+ * Resolver (the Resolver-Node relationship)
+ *   - has one `addr` record for a Node
```
Is there any guaranteed relationship in the subgraph between the `addr` field being defined and the presence of a particular `coinTypes` key? No matter if that's a yes or no, this would be a nice invariant / expectation to document.

lightwalker-eth commented on apps/ensindexer/.env.local.example:
```diff
@@ -87,11 +87,18 @@ ENSADMIN_URL=https://admin.ensnode.io
+#
+# NOTE: Setting this to `true` results in the resulting indexed data being a strict _superset_ of
+# subgraph data. For exact subgraph data compatibility, set this to `false`.
 INDEX_RESOLVER_VALUES=true
```
I believe we need to rename this. Could you please search all files for other places where this new env var should be renamed or introduced? For example, what about our docker configs, etc? Appreciate your advice.

Could you please also share advice for how we should track the need to update environment variable configurations as part of the deployment of this new release? For example, we should explicitly add configurations to this environment variable across all of our ENSNode instances. Perhaps we should add a GitHub issue for this and assign it to Sebastian and/or Tko? I'm worried we will forget about this important follow up action item when we deploy the next release.

lightwalker-eth commented on apps/ensindexer/.env.local.example:
```diff
@@ -87,11 +87,18 @@ ENSADMIN_URL=https://admin.ensnode.io
+# record values in their applications. Features are planned in the ENSNode roadmap that will provide
+# safe use of indexed resolver record values (in appropriate contexts).
+#
+# NOTE: Setting this to `true` results in the resulting indexed data being a strict _superset_ of
```
```suggestion
# NOTE: Setting this to `true` results in indexed data being a strict _superset_ of
```

lightwalker-eth reviewed (changes requested):
@shrugs Great updates 😄 Reviewed and shared a few suggestions :+1:

lightwalker-eth commented on terraform/modules/ensindexer/main.tf:
```diff
@@ -99,6 +99,10 @@ resource "railway_variable_collection" "ensindexer" {
       value = var.heal_reverse_addresses
     },
+    {
+      name  = "INDEX_ADDITION_RESOLVER_RECORDS"
```
```suggestion
      name  = "INDEX_ADDITIONAL_RESOLVER_RECORDS"
```

lightwalker-eth commented on terraform/modules/ensindexer/main.tf:
```diff
@@ -180,6 +184,10 @@ resource "railway_variable_collection" "ensindexer_api" {
       value = var.heal_reverse_addresses
     },
+    {
+      name  = "INDEX_ADDITION_RESOLVER_RECORDS"
```
```suggestion
      name  = "INDEX_ADDITIONAL_RESOLVER_RECORDS"
```

lightwalker-eth commented on apps/ensindexer/.env.local.example:
```diff
@@ -87,18 +87,16 @@ ENSADMIN_URL=https://admin.ensnode.io
-# record values in their applications. Features are planned in the ENSNode roadmap that will provide
-# safe use of indexed resolver record values (in appropriate contexts).
+# WARNING: It is unsafe to naively assume that indexed resolver record values are equivalent to the
+# result of performing dynamic forward resolution via the ENS protocol, as retrieving records values
```
```suggestion
# result of performing dynamic forward resolution via the ENS protocol, as naively retrieving resolver records
```

lightwalker-eth commented on apps/ensindexer/.env.local.example:
```diff
@@ -87,18 +87,16 @@ ENSADMIN_URL=https://admin.ensnode.io
-# safe use of indexed resolver record values (in appropriate contexts).
+# WARNING: It is unsafe to naively assume that indexed resolver record values are equivalent to the
+# result of performing dynamic forward resolution via the ENS protocol, as retrieving records values
+# from an index is not ENSIP-10 nor CCIP-Read compliant. At this time we do _not_ recommend anyone
```
```suggestion
# from indexed may not be ENSIP-10 nor CCIP-Read compliant. At this time we do _not_ recommend anyone
```

lightwalker-eth commented on packages/ensnode-schema/src/resolver-records.schema.ts:
```diff
@@ -36,47 +36,23 @@
  *   + has many `addressRecords` for a Node (by coinType) (see ENSIP-9)
- *     - ResolverAddressRecord indexes `coinType` and `address`
+ *     - see ResolverAddressRecord below
+ *     - NOTE: _does_ represent the implcit coinType of 60 for the `addr` record, if set
```
```suggestion
 *     - NOTE: _does_ represent the implicit coinType of 60 for the `addr` record, if set
```

lightwalker-eth reviewed (approved):
@shrugs Great work on all of this! 🚀 Major new capability 💪

lightwalker-eth reviewed (approved):
Sharing a few more suggestions :+1:

shrugs commented:
i was able to trivially filter out the ext_ prefixed tables/relations from the schema before passing to the generated apis so that seems to be the simplest approach here for sure
